### PR TITLE
[DO NOT MERGE] feature/px4 merge

### PR DESCRIPTION
**Pull request by gbmhunter**

### CODE REVIEW

**Comment by ffurrer** [comment]
ok to test.

**Comment by helenol** [comment]
test this please -- @gbmhunter, @burrimi and I fixed the protobuf compile issues on ubuntu Jenkins, but it still doesn't compile for code reasons. ;)

**Comment by gbmhunter** [comment]
@helenol Thank you for fixing the protobuf issue!

**Comment by helenol** [comment]
Looks like it may be different versions of gazebo actually that are causing the issue on 14.04, but rotors still needs to be able to work on the old one :/ But you can use calls to the older API with defines in the code? This is how the standard gazebo plugins do this.

**Comment by gbmhunter** [comment]
@helenol Yes I will have a look into this.

**Comment by gbmhunter** [comment]
@helenol @ffurrer Have a look at [here](https://bitbucket.org/osrf/gazebo/src/d115a95b5c947563ba641917895d0efbbbf82bc8/gazebo/util/system.hh?at=gazebo5_5.1.0&fileviewer=file-view-default), it seems to suggest that `system.hh` is in Gazebo 5. 

To further test, I installed Gazebo 5 and it seemed to compile correctly.

**Comment by gbmhunter** [comment]
Implemented in 960064382eeadb46bb29b7e2c41ba71bad65d61d.

File: rotors_gazebo_plugins/src/gazebo_mavlink_interface.cpp [hunk 1]
```diff
@@ -4,6 +4,7 @@
  * Copyright 2015 Mina Kamel, ASL, ETH Zurich, Switzerland
  * Copyright 2015 Janosch Nikolic, ASL, ETH Zurich, Switzerland
  * Copyright 2015 Markus Achtelik, ASL, ETH Zurich, Switzerland
+ * Copyright 2016 Geoffrey Hunter <gbmhunter@gmail.com>
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_mavlink_interface.cpp]: @LorenzMeier can you have a look at the mavlink side? I never used it.

File: rotors_gazebo_plugins/src/gazebo_motor_model.cpp [hunk 2]
```diff
@@ -104,47 +117,140 @@ void GazeboMotorModel::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf) {
...
+  //connect_ros_to_gazebo_topic_msg.set_gazebo_namespace(namespace_);
+  connect_ros_to_gazebo_topic_msg.set_gazebo_topic("~/" + namespace_ + "/" + command_sub_topic_);
+  connect_ros_to_gazebo_topic_msg.set_msgtype(gz_std_msgs::ConnectRosToGazeboTopic::COMMAND_MOTOR_SPEED);
+  gz_connect_ros_to_gazebo_topic_pub->Publish(connect_ros_to_gazebo_topic_msg, true);
+
+  // ============================================ //
+  // ==== WIND SPEED MSG SETUP (ROS->GAZEBO) ==== //
+  // ============================================ //
+
+  wind_speed_sub_ = node_handle_->Subscribe("~/" + namespace_ + "/" + wind_speed_sub_topic_, &GazeboMotorModel::WindSpeedCallback, this);
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_motor_model.cpp]: do we need windspeed here? I think we have a wind_speed_plugin that should connect to the ros topics.

File: rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp [hunk 3]
```diff
@@ -236,47 +257,69 @@ void GazeboOdometryPlugin::OnUpdate(const common::UpdateInfo& _info) {
...
-    odometry.pose.pose.orientation.z = gazebo_pose.rot.z;
-    odometry.twist.twist.linear.x = gazebo_linear_velocity.x;
-    odometry.twist.twist.linear.y = gazebo_linear_velocity.y;
-    odometry.twist.twist.linear.z = gazebo_linear_velocity.z;
-    odometry.twist.twist.angular.x = gazebo_angular_velocity.x;
-    odometry.twist.twist.angular.y = gazebo_angular_velocity.y;
-    odometry.twist.twist.angular.z = gazebo_angular_velocity.z;
+
+    gz_geometry_msgs::Odometry odometry;
+    odometry.mutable_header()->set_frame_id(parent_frame_id_);
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp]: add issue. move these conversions to a common header.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 4]
```diff
@@ -0,0 +1,938 @@
...
+  //==============================================//
+
+  if (_sdf->HasElement("robotNamespace"))
+    namespace_ = _sdf->GetElement("robotNamespace")->Get<std::string>();
+  else
+    gzerr << "Please specify a robotNamespace.\n";
+  gzdbg << "namespace_ = \"" << namespace_ << "\"." << std::endl;
+
+  /// @todo Fix this hack!!! Should not need this namespace set to firefly.
+  namespace_ = "firefly";
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: We need to make sure it works with different types of mavs.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 5]
```diff
@@ -0,0 +1,938 @@
...
+
+void GazeboRosInterfacePlugin::OnUpdate(const common::UpdateInfo& _info) {
+  // Do nothing
+  // This plugins actions are all executed through message callbacks.
+}
+
+
+/// \brief      A helper class that provides storage for additional parameters that are inserted
+///             into the callback.
+template <typename M>
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: what is M?

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 6]
```diff
@@ -0,0 +1,938 @@
...
+
+GazeboRosInterfacePlugin::GazeboRosInterfacePlugin()
+    : WorldPlugin(),
+      gz_node_handle_(0),
+      ros_node_handle_(0),
+      ros_actuators_msg_(new mav_msgs::Actuators) { }
+
+
+GazeboRosInterfacePlugin::~GazeboRosInterfacePlugin() {
+  event::Events::DisconnectWorldUpdateBegin(updateConnection_);
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: `delete ros_node_handle_`;

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 7]
```diff
@@ -0,0 +1,938 @@
...
+  if(kPrintOnMsgCallback) {
+    gzdbg << __FUNCTION__ << "() called." << std::endl;
+  }
+
+  static std::vector<ros::Subscriber> ros_subscribers;
+
+  switch(gz_connect_ros_to_gazebo_topic_msg->msgtype()) {
+    case gz_std_msgs::ConnectRosToGazeboTopic::ACTUATORS: {
+
+      // Create Gazebo publisher
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: remove dead code

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 8]
```diff
@@ -0,0 +1,938 @@
...
+  ros_publisher.publish(ros_imu_msg_);
+
+}
+
+void GazeboRosInterfacePlugin::GzJointStateMsgCallback(GzJointStateMsgPtr& gz_joint_state_msg, ros::Publisher ros_publisher) {
+
+  // ============================================ //
+  // =================== HEADER ================= //
+  // ============================================ //
+  ros_joint_state_msg_.header.stamp.sec = gz_joint_state_msg->header().stamp().sec();
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: create helper function for header conversion.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp [hunk 9]
```diff
@@ -0,0 +1,938 @@
...
+    std::string rosTopicName,
+    transport::NodePtr gz_node_handle) {
+
+  // One map will be created for each type M
+  static std::map< std::string, ConnectHelperStorage<M> > callback_map;
+
+  // Create ROS publisher
+  ros::Publisher ros_publisher = ros_node_handle_->advertise<N>(rosTopicName, 1);
+
+  /// \todo Handle collision error
```
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: add a check if this topic name is already in callback_map!

File: rotors_gazebo/scripts/xacro.py [hunk 10]
```diff
@@ -0,0 +1,709 @@
+# Copyright (c) 2013, Willow Garage, Inc.
```
Comment (gbmhunter) [inline rotors_gazebo/scripts/xacro.py]: Leave as is for now.

File: rotors_gazebo/scripts/substitution_args.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
+# Software License Agreement (BSD License)
```
Comment (gbmhunter) [inline rotors_gazebo/scripts/substitution_args.py]: Leave as is for now.

File: rotors_gazebo/scripts/names.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
+# Software License Agreement (BSD License)
```
Comment (gbmhunter) [inline rotors_gazebo/scripts/names.py]: Leave as is for now.

File: rotors_gazebo_plugins/CMakeLists.txt [hunk 13]
```diff
@@ -1,142 +1,422 @@
+# Optional arguments to be passed into file
+# ADDITIONAL_INCLUDE_DIRS           string  Additional include directories to add to every build target (PX4 uses this).
+# BUILD_MAVLINK_INTERFACE_PLUGIN    bool    Build mavlink_interface_plugin (requires mav dependency).
+# BUILD_OCTOMAP_PLUGIN              bool    Build the optical map plugin (requires Octomap).
+# BUILD_OPTICAL_FLOW_PLUGIN         bool    Build the optical flow plugin (requires OpenCV).
+# MAVLINK_HEADER_DIR                string  Location of MAVLink header files. If not provided, this CMakeLists.txt file will
+#                                               search the default locations (e.g. ROS) for them. This variable is only required
+#                                               if BUILD_MAVLINK_INTERFACE_PLUGIN=TRUE.
+# NO_ROS                            bool    Build without any ROS dependencies.
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/CMakeLists.txt]: Added defaults (for booleans, the strings already have their default cases handled).

File: rotors_gazebo_plugins/CMakeLists.txt [hunk 14]
```diff
@@ -1,142 +1,422 @@
...
+
+if (BUILD_MAVLINK_INTERFACE_PLUGIN)
+
+  # We need the MAVLink headers.
+  set(MAVLINK_HEADERS_FOUND FALSE)
+  
+  # First, check to see if MAVLink headers were passed in as variable
+  if(NOT ${MAVLINK_HEADER_DIR} STREQUAL "")
+  
+    message(STATUS "MAVLINK_HEADER_DIR provided as '${MAVLINK_HEADER_DIR}'.")
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/CMakeLists.txt]: Yes that would be tidier, as it would separate concerns. Can we leave for now though in the interests of getting things merged? Can make issue if you want.

File: rotors_gazebo_plugins/CMakeLists.txt [hunk 15]
```diff
@@ -1,142 +1,422 @@
...
+    include_directories(${MAVLINK_HEADER_DIR})
+    set(MAVLINK_HEADERS_FOUND TRUE)
+    
+  else()
+  
+    message(STATUS "MAVLINK_HEADER_DIR not provided, falling back to looking at default paths...")
+    # If ROS indigo is installed, we should be able to find them at the path below
+    if(EXISTS /opt/ros/indigo/include/mavlink/v1.0/)
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/CMakeLists.txt]: Would a findMavLink.cmake script, with knowledge about all current ROS versions, suffice? Perhaps it could even look to find `/opt/ros/.*/include/mavlink/v1.0/` where `.*` or similar is a Glob pattern to match any ROS version.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_gimbal_controller_plugin.h [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+
+    public: virtual void Load(physics::ModelPtr _model, sdf::ElementPtr _sdf);
+
+    public: virtual void Init();
+
+    private: void OnUpdate();
+
+#if GAZEBO_MAJOR_VERSION >= 7 && GAZEBO_MINOR_VERSION >= 4
+    /// only gazebo 7.4 and above support Any
+    private: void OnPitchStringMsg(ConstAnyPtr &_msg);
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_gimbal_controller_plugin.h]: Moved to `external`.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_gimbal_controller_plugin.h [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    private: double NormalizeAbout(double _angle, double _reference);
+
+    /// \TODO something to move into Angle class
+    /// \brief returns shortest angular distance from _from to _to
+    /// \param[in] _from starting anglular position
+    /// \param[in] _to end angular position
+    /// \return distance traveled from starting to end angular positions
+    private: double ShortestAngularDistance(double _from, double _to);
+
+    private: sdf::ElementPtr sdf;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_gimbal_controller_plugin.h]: Moved to `external`.

File: rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp [hunk 18]
```diff
@@ -211,8 +260,18 @@ void GazeboImuPlugin::addNoise(Eigen::Vector3d* linear_acceleration,
...
-// This gets called by the world update start event.
+
 void GazeboImuPlugin::OnUpdate(const common::UpdateInfo& _info) {
+
+  if(kPrintOnUpdates) {
+    gzdbg << __FUNCTION__ << "() called." << std::endl;
+  }
+
+  if(!pubs_and_subs_created_) {
+    CreatePubsAndSubs();
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp]: This was a problem I had a while back. I used to have publisher/subscriber creation in `Load()` but I ran into a problem in where the ROS interface plugin would then sometimes miss "connection" messages. I think that because the plugin order is somewhat undefined, if say, the IMU plugin's `Load()` gets called first, and publishes a "connect" message to the ROS interface plugin, the ROS interface plugin will miss this if it has not yet be initialised. 

I don't think messages queue if there is no subscriber, I think they just get dropped. My solution was to put the pub/sub creation into the `Run()` method, as the ROS interface plugin is guaranteed to be running at this point. Can you think of a better solution?

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h [hunk 19]
```diff
@@ -18,90 +18,196 @@
...
+#include "gazebo/transport/transport.hh"
+#include "gazebo/msgs/msgs.hh"
+#include "common/mavlink.h"     // Either provided by ROS or as CMake argument MAVLINK_HEADER_DIR
+#include "common/mavlink_msg_hil_actuator_controls.h"
 
-#include "rotors_gazebo_plugins/common.h"
+#include "common.h"
+//#include "mavlink/v1.0/common/mavlink.h"
+
+#include "CommandMotorSpeed.pb.h"
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h]: I think `MotorSpeed` was what PX4 used and `CommandMotorSpeed` was what ASL used. I have removed `MotorSpeed` and it's unused inclusion in the MAVLink interface plugin.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h [hunk 20]
```diff
@@ -18,90 +18,196 @@
...
+  boost::thread callback_queue_thread_;
+  void QueueThread();
+  void ImuCallback(ImuPtr& imu_msg);
+  void LidarCallback(LidarPtr& lidar_msg);
+  void OpticalFlowCallback(OpticalFlowPtr& opticalFlow_msg);
+  void send_mavlink_message(const uint8_t msgid, const void *msg, uint8_t component_ID);
+  void handle_message(mavlink_message_t *msg);
+  void pollForMAVLinkMessages(double _dt, uint32_t _timeoutMs);
+
+  static const unsigned n_out_max = 16;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h]: Done.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h [hunk 21]
```diff
@@ -18,90 +18,196 @@
...
+
+  unsigned _rotor_count;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h]: Done.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h [hunk 22]
```diff
@@ -18,90 +18,196 @@
...
   
   common::Time last_time_;
   common::Time last_gps_time_;
+  common::Time last_actuator_time_;
   double gps_update_interval_;
-  int rotor_count_;
-
-  mavlink_hil_sensor_t hil_sensor_msg_;
-  mavlink_hil_gps_t hil_gps_msg_;
+  double lat_rad;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h]: Done.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h [hunk 23]
```diff
@@ -18,90 +18,196 @@
...
+  double lon_rad;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h]: Done.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h [hunk 24]
```diff
@@ -18,90 +18,196 @@
...
 
-  double ref_lat_;
-  double ref_lon_;
-  double ref_alt_;
-};
-}
+  std::default_random_engine random_generator_;
+  std::normal_distribution<float> standard_normal_distribution_;
+
+  int _fd;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_mavlink_interface.h]: Done.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_motor_model.h [hunk 25]
```diff
@@ -30,26 +30,30 @@
...
 const static int CCW = 1;
 const static int CW = -1;
-}
+} // namespace turning_direction
 
 namespace gazebo {
+
 // Default values
 static const std::string kDefaultCommandSubTopic = "gazebo/command/motor_speed";
 static const std::string kDefaultWindSpeedSubTopic = "gazebo/wind_speed";
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_motor_model.h]: `mav_comm` stuff should still work (I download the repo on the PX4 build also). I have now used constants from `mav_comm`.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_ros_interface_plugin.h [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+  void InitializeParams();
+  void Publish();
+
+ protected:
+
+
+  void Load(physics::WorldPtr _world, sdf::ElementPtr _sdf);
+
+  /// @brief  	This gets called by the world update start event.
+  /// @details	Calculates IMU parameters and then publishes one IMU message.
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_ros_interface_plugin.h]: Fixed.

File: rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_ros_interface_plugin.h [hunk 27]
```diff
@@ -0,0 +1,272 @@
...
+  void OnUpdate(const common::UpdateInfo&);
+
+ private:
+
+  /// \brief  Provides a way for GzConnectGazeboToRosTopicMsgCallback() to connect a Gazebo subscriber to
+  ///         a ROS publisher.
+  /// \details
+  ///   M is the type of the message that will be subscribed to the Gazebo framework.
+  ///   N is the type of the message published to the ROS framework
+  template <typename M, typename N>
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/include/rotors_gazebo_plugins/gazebo_ros_interface_plugin.h]: Done.

File: rotors_gazebo_plugins/msgs/MotorSpeed.proto [hunk 28]
```diff
@@ -0,0 +1,7 @@
+syntax = "proto2";
+package mav_msgs.msgs;
+
+message MotorSpeed
+{
+  repeated float motor_speed = 1 [packed=true];
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/msgs/MotorSpeed.proto]: Message type deleted, using `CommandMotorSpeed` only.

File: rotors_gazebo_plugins/msgs/CommandMotorSpeed.proto [hunk 29]
```diff
@@ -0,0 +1,6 @@
+syntax = "proto2";
+package gz_mav_msgs;
+
+message CommandMotorSpeed {
+  repeated float motor_speed = 1 [packed=true];
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/msgs/CommandMotorSpeed.proto]: Why not `float`? Are motor speed commands always `integers`? Or are you asking why is it not higher precision (i.e. `double`)?

File: rotors_gazebo_plugins/msgs/TwistStamped.proto [hunk 30]
```diff
@@ -0,0 +1,30 @@
...
+      required uint32 nsec = 2;
+    }
+    required string frame_id = 1;
+    required Stamp stamp = 2;
+  }
+  required Header header = 1;
+
+  message Twist {
+    message Linear {
+      required double x = 1;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/msgs/TwistStamped.proto]: Fixed.

File: rotors_gazebo_plugins/msgs/TwistStamped.proto [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+      required double y = 2;
+      required double z = 3;
+    }
+    required Linear linear = 1;
+  }
+  required Twist twist = 2;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/msgs/TwistStamped.proto]: Fixed.

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 32]
```diff
@@ -69,8 +83,10 @@ void GazeboGpsPlugin::Load(sensors::SensorPtr _sensor, sdf::ElementPtr _sdf) {
   double hor_vel_std_dev;
   double ver_vel_std_dev;
 
-  getSdfParam<std::string>(_sdf, "gpsTopic", gps_topic_,
-                           mav_msgs::default_topics::GPS);
+//  getSdfParam<std::string>(_sdf, "gpsTopic", gps_topic_,
+//                           mav_msgs::default_topics::GPS);
+  getSdfParam<std::string>(_sdf, "gpsTopic", gps_topic_, "");
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: Added from `mav_msgs`.

File: rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp [hunk 33]
```diff
@@ -86,29 +102,61 @@ void GazeboGpsPlugin::Load(sensors::SensorPtr _sensor, sdf::ElementPtr _sdf) {
...
 
 void GazeboGpsPlugin::OnUpdate() {
+
+  if(kPrintOnUpdates) {
+    gzdbg << __FUNCTION__ << "() called." << std::endl;
+  }
+
+  if(!pubs_and_subs_created_) {
+    CreatePubsAndSubs();
+    pubs_and_subs_created_ = true;
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_gps_plugin.cpp]: Same reason as above.

File: rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp [hunk 34]
```diff
@@ -18,44 +19,61 @@
  * limitations under the License.
  */
 
+// MODULE HEADER
 #include "rotors_gazebo_plugins/gazebo_imu_plugin.h"
 
+// USER HEADERS
+#include "ConnectGazeboToRosTopic.pb.h"
+
+// SYSTEM LIBS
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp]: The Google style guide suggests:
related header
system
3rd party
user

I have changed it to follow this.

File: rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp [hunk 35]
```diff
@@ -243,33 +302,73 @@ void GazeboImuPlugin::OnUpdate(const common::UpdateInfo& _info) {
...
+  // not raw IMU data but rather a calculation (and could confuse users).
+  // However, the orientation is now set as it is used by PX4.
+  /*gazebo::msgs::Quaternion* orientation = new gazebo::msgs::Quaternion();
+  orientation->set_x(0);
+  orientation->set_y(0);
+  orientation->set_z(0);
+  orientation->set_w(1);
+  imu_message_.set_allocated_orientation(orientation);*/
+
+  gazebo::msgs::Quaternion* orientation = new gazebo::msgs::Quaternion();
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp]: Done.

File: rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp [hunk 36]
```diff
@@ -99,14 +110,46 @@ void GazeboMagnetometerPlugin::Load(physics::ModelPtr _model, sdf::ElementPtr _s
...
+
 }
 
 void GazeboMagnetometerPlugin::OnUpdate(const common::UpdateInfo& _info) {
+
+  if(kPrintOnUpdates) {
+    gzdbg << __FUNCTION__ << "() called." << std::endl;
+  }
+
+  if(!pubs_and_subs_created_) {
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp]: Same as above.

File: rotors_gazebo_plugins/src/gazebo_mavlink_interface.cpp [hunk 37]
```diff
@@ -18,226 +19,1019 @@
...
-GazeboMavlinkInterface::GazeboMavlinkInterface()
-    : ModelPlugin(),
-      node_handle_(0) {}
+// Set global reference point
+// Zurich Irchel Park: 47.397742, 8.545594, 488m
+// Seattle downtown (15 deg declination): 47.592182, -122.316031, 86m
+// Moscow downtown: 55.753395, 37.625427, 155m
+
+// Zurich Irchel Park
+static const double lat_zurich = 47.397742 * M_PI / 180;  // rad
```
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_mavlink_interface.cpp]: Fixed style. Haven't moved to header, not sure whether this is the best thing to do?

File: rotors_gazebo_plugins/src/gazebo_motor_model.cpp (hunk 2 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_motor_model.cpp]: Not sure. I have not changed anything here, this was as is, I'm thinking a change is out of scope of this pull-request.

File: rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp (hunk 3 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp]: Sorry, not quite sure what is require here? Do you want me to add an issue about:

> // TODO: covariance scaling, according to the intensity values could be implemented here.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 6 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Done.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 4 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: I have fixed this. A namespace variable is no longer needed for the ROS interface plugin. Haven't checked it works with other MAVs yet.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 5 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Fixed.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 9 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Done.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 7 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Done.

File: rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp (hunk 8 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp]: Done.

File: rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp (hunk 3 above)
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp]: sorry was a bit unclear. I thought of a function that converts a gazebo header to a ros header since all headers have the same structure and you could safe a lot of code duplication

File: rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp (hunk 18 above)
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp]: can you document this somewhere? otherwise this knowledge is lost. maybe in the wiki?

File: rotors_gazebo_plugins/msgs/CommandMotorSpeed.proto (hunk 29 above)
Comment (burrimi) [inline rotors_gazebo_plugins/msgs/CommandMotorSpeed.proto]: was thinking of doubles, but I guess this is fine

File: rotors_gazebo_plugins/src/gazebo_motor_model.cpp (hunk 2 above)
Comment (burrimi) [inline rotors_gazebo_plugins/src/gazebo_motor_model.cpp]: ok can you add a TODO(gbmhunter): do we need this?

File: rotors_gazebo_plugins/CMakeLists.txt (hunk 15 above)
Comment (burrimi) [inline rotors_gazebo_plugins/CMakeLists.txt]: yes and the most common path for people who install mavlink as a system library.

File: rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp (hunk 18 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_imu_plugin.cpp]: Yes, have documented at https://github.com/ethz-asl/rotors_simulator/wiki/ROS-Interface-Plugin.

File: rotors_gazebo_plugins/src/gazebo_motor_model.cpp (hunk 2 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_motor_model.cpp]: Done.

File: rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp (hunk 3 above)
Comment (gbmhunter) [inline rotors_gazebo_plugins/src/gazebo_odometry_plugin.cpp]: Ah right, yes there is now a header helper method which can convert a Gazebo header to a ROS header, and another for the reverse direction.

**Comment by burrimi** [comment]
test this please
